_code/insitucor-custom-module/InSituCor_Module.R

```diff
@@ -0,0 +1,398 @@
+
+counts <- study$somas$RNA$X$members$counts$to_matrix()
+print(str(counts))
+counts <- Matrix::t(counts)
+norm <- sweep(counts, 1, pmax(Matrix::rowSums(counts), 20), "/") * 500
+rm(counts)
+
+if (verbose) {
+  print("reached2.4")
+}
+
+# align norm to metadata:
+annotcellIDs <- study$somas$RNA$obs$to_dataframe("cell_ID")
+annotcellIDs <- annotcellIDs$cell_ID
+
+m <- match(annotcellIDs, rownames(norm))
+norm <- norm[m, ]
+
+# using slide ID as a stand-in for tissue, which won't generally be available
+tissuevec <- study$somas$RNA$obs$to_dataframe("slide_ID_numeric")  
+tissuevec <- tissuevec$slide_ID_numeric
+
+print(table(tissuevec))
+print(head(tissuevec))
+
+# get xy:
+xy <- as.matrix(cbind(study$somas$RNA$obs$to_dataframe("x_slide_mm"),
+                      study$somas$RNA$obs$to_dataframe("y_slide_mm")))
+
+if (verbose) {
+  print("reached3")
+  print(head(xy))
+}
+
+
+
+#### get cell type: ---------------------------------
+
+# (dev note: this doesn't yet allow for the use of refineClusters, or for any
+#  custom cell type assignments. This logic should eventually be expanded to do so.)
+
+celltype <- match.arg(celltype, c("Insitutype", "Leiden"))
+obs_colns <- study$somas$RNA$obs$attrnames()
+if(celltype == "Insitutype"){
+  celltypecolumn <- obs_colns[max(which(grepl("nbclust", obs_colns) & grepl("clusters", obs_colns)))] #(column on the right is the latest)
+}
+if (celltype == "Leiden") {
+  celltypecolumn <- obs_colns[max(which(grepl("nn", obs_colns)))]
+  #celltype <- grep("^nn_.*_cluster_cluster_", obs_colns, value = T)[1]
+}
+if(celltypecolumn < 0) {
+  stop(paste0("couldn't find the cell type column for ", celltype))
+}
+celltypevec <- study$somas$RNA$obs$to_dataframe()[[celltypecolumn]]
+
+if (verbose) {
+  print("reached4")
+  print(celltypecolumn)
+  print(head(celltypevec))
+  print(str(celltypevec))
+  print(head(study$somas$RNA$obs$to_dataframe()))
+  print(str(study$somas$RNA$obs$to_dataframe()))
+  print(head(annotcellIDs))
+  print(head(tissuevec))
+  print(head(celltypevec))
+  print(head(xy))
+  print(norm[1:4,1:4])
+}
+
+
+# subset to complete data:
+iscomplete <- as.logical((Matrix::rowSums(is.na(norm)) == 0) * !is.na(celltypevec) * !is.na(tissuevec) * (rowSums(is.na(xy)) == 0))
+norm <- norm[iscomplete, ]
+celltypevec <- celltypevec[iscomplete]
+tissuevec <- tissuevec[iscomplete]
+xy <- xy[iscomplete, ]
+
+print(table(iscomplete))
+print(dim(norm))
+print(length(celltypevec))
+
+if (verbose) {
+  print("reached5")
+}
+
+#### run InSituCor ---------------------------------------
+# prep conditionon:
+conditionon <-  data.frame(celltype = celltypevec)
+rownames(conditionon) <- rownames(norm)
+
+if (verbose) {
+  print("reached6")
+}
+
+# remove genes with no information:
+propnonzero <- colMeans(norm > 0)
+
+# run insitucor:
+res <- insitucor(       
+  counts = norm[, propnonzero > 0.005], 
+  conditionon = conditionon,
+  celltype = celltypevec,
+  neighbors = NULL,
+  xy = xy,
+  k = 50,
```

Comment:
Those should be exposed as input variables for the Custom Module if possible.

Reply:
we made a decision to keep the user's life simple and not expose more than the absolute necessary arguments. 

Added comments to the README saying you can edit the arguments within the script.